qokit/qaoa_circuit_utils.py

```diff
@@ -0,0 +1,291 @@
+###############################################################################
+# // SPDX-License-Identifier: Apache-2.0
+# // Copyright : JP Morgan Chase & Co
+###############################################################################
+# QAOA circuit for S_k
+
+import numpy as np
+from qiskit import QuantumCircuit, QuantumRegister, ClassicalRegister
+from qiskit.circuit import ParameterVector
+from typing import Sequence
+
+
+def append_zz_term(qc, q1, q2, gamma):
+    qc.rzz(gamma, q1, q2)
+
+
+def append_z_prod_term(qc: QuantumCircuit, term: Sequence, gamma: float) -> None:
+    """Appends a  multi-body Pauli-Z interaction acting on qubits whose indices
+    correspond to those in 'term'.
+
+    Parameters:
+        qc: QuantumCircuit
+        term: iterable
+            ordered iterable containing qubit indices to apply Pauli-Z interaction to
+        gamma: float
+            evolution time for interaction
+
+    """
+    # term_weight, term = term
+    term_weight = len(term)
+    assert all(term[i] < term[i + 1] for i in range(len(term) - 1))
+    if term_weight == 4:
+        # in labs, four-body terms appear two times more than two-body
+        # there is also a global scaling factor of 2 for all terms (four and two), which is ignored here
+        _gamma = 2 * gamma
+        qc.cx(term[0], term[1])
+        qc.cx(term[3], term[2])
+        qc.rzz(2 * _gamma, term[1], term[2])
+        qc.cx(term[3], term[2])
+        qc.cx(term[0], term[1])
+    elif term_weight == 2:
+        qc.rzz(2 * gamma, term[0], term[1])
+    else:
+        # fallback to general case
+        target = term[-1]
+        for control in term[:-1]:
+            qc.cx(control, target)
+        qc.rz(2 * gamma, target)
+        for control in term[:-1]:
+            qc.cx(control, target)
+
+
+def append_cost_operator_circuit(qc, terms, gamma):
+    for term, (i, j) in terms:
```

Comment:
Would be more clear if we used variable name `coeff` instead of `term`,  i.e. `for coeff, (i, j) in terms:`

Reply:
Done!